### ISSUE TITLE
Fix streaming from network drives

### ISSUE THREAD

**Issue by oxan** [MPExtended/MPExtended#59]
I think we need to run the transcoder in the impersonation context.

**Comment by oxan** [comment]
This is troubling:
- Starting a process as a different user (CreateUserWithLogonW) doesn't work as the LocalSystem account (access denied)
- The other APIs which allow that aren't wrapped in .NET
- We can't read the file ourselves as for quite some container formats (MKV, AVI) the transcoder has problems with that

**Comment by Ollienator** [comment]
Hi Oxan,
could you check out commit efa9ba6 in my fork?
It should fix this issue.  As you said, it's not trivial.

**Comment by oxan** [comment]
Wow, that's awesome! I didn't expect it to be so much work, so thanks again!

I've one question about the code though: why did you opt to remove MediaSource.GetImpersonator()? The idea (which, I've to admit, wasn't documented at all) behind that was that we could always use the same code in places where impersonation might be needed, and it switched impersonation on only when it was needed. That is, we could write code like this:

```
using (var impersonator = source.GetImpersonator()) {
   // do some stuff with the file
}
```

Which resulted in impersonation being enabled only when the file needed it. I'm not really opposed to removing it, but I'm curious as to whether there was a fundamental problem with it or you just didn't get the idea or didn't like it ;)

In any case, the code is pretty awesome. I've a few minor stylistic comments which I've added inline to your code, but that's nothing serious. I'll run some more rigorous tests tomorrow, and if that doesn't reveal any serious problems, I'll merge it. 

Again, thank you very much!

**Comment by oxan** [comment]
Argh, GitHub is stupid, it only shows the inline comments I made if you look up the commit in your repository: https://github.com/Ollienator/MPExtended/commit/efa9ba6

**Comment by Ollienator** [comment]
No there weren't any deeper thoughts involved.  I just saw mixed "source.GetImpersonator()" and "new NetworkShareImperonator()" calls in different places, so I thought I'd make it more consistent (as there are places when there is no "source" object available to use GetImpersonator(), if I remember correctly).

**Comment by Ollienator** [comment]
I added your suggestions in my latest commit.  Tried streaming TV, music and movies and it seems to work fine.  More testing needed, obviously.
One thing I have noticed though:  Sometimes (and so far I haven't found a way to reproduce this) there is no impersonation done when there should be - the stream plays fine for a file on a network share that the LocalSystem account shouldn't have access to when running the transcoder.  Well, it works, so I guess this is ok anyway.
